### PR TITLE
added bedrock distrubution

### PR DESCRIPTION
**Pull request by faisalnaveedRNS**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/pylons/keeper/distribution.go [hunk 1]
```diff
@@ -67,35 +74,109 @@
...
+				amountForAddr := availableAmount.Mul(percentage)
+				if amountForAddr.IsPositive() {
+					// only add strictly positive amounts
+					totalAmountsForAddr = totalAmountsForAddr.Add(sdk.NewCoin(coin.Denom, amountForAddr.RoundInt()))
+				}
+			}
+			if !totalAmountsForAddr.Empty() {
+				delegatorsRewards[addr] = totalAmountsForAddr
+			}
+		}
```
[the remark this answers is not in the document]
Comment (faisalnaveedRNS) [reply x/pylons/keeper/distribution.go]: resolved

File: x/pylons/keeper/distribution.go [hunk 2]
```diff
@@ -67,35 +74,109 @@
...
```
[the remark this answers is not in the document]
Comment (faisalnaveedRNS) [reply x/pylons/keeper/distribution.go]: resolved

File: x/pylons/keeper/hooks_test.go [hunk 3]
```diff
@@ -267,8 +272,201 @@ func (suite *IntegrationTestSuite) TestAfterEpochEndNoDeligators() {
...
+					coins: sdk.Coins{
+						sdk.Coin{Denom: types.StakingCoinDenom, Amount: sdk.NewInt(100)},
+					},
+					expectedCoins: sdk.Coins{
+						sdk.Coin{Denom: types.PylonsCoinDenom, Amount: sdk.NewInt(3)},
+					},
+				},
+			},
+			delegatorBalance:        1,
+			updatedDelegatorBalance: 2,
```
Comment (MikeSofaer) [inline x/pylons/keeper/hooks_test.go:384]: Seems like the numbers here only add to 96?

**Review by MikeSofaer — CHANGES_REQUESTED** [review]
I would like to make sure validators are called validators and not delegators, delegators are holders, not validators.

Also the math isn't super clear, and in the multi-holder test the two holders should have different quantities of Bedrock and get different rewards.

File: x/pylons/keeper/hooks_test.go (hunk 3 above)
Comment (faisalnaveedRNS) [reply x/pylons/keeper/hooks_test.go:384]: need more information

**Comment by faisalnaveedRNS** [comment]
> 

updated test cases @MikeSofaer

File: x/pylons/keeper/hooks_test.go (hunk 3 above)
Comment (faisalnaveedRNS) [reply x/pylons/keeper/hooks_test.go]: I am resolving this issue for now. let me know if any thing need to change,

**Review by MikeSofaer — APPROVED** [review]

**Pull request merged**